fastlane/Fastfile

```diff
@@ -6,13 +6,13 @@ platform :android do
 
   desc "Pull Request verification"
   lane :pull_request_verification do
-    gradle(task: "clean --no-daemon")
-    gradle(task: "detekt --no-daemon")
-    gradle(task: "lintDebug --no-daemon")
-    gradle(task: "jacocoDebugTestReport --no-daemon")
-    gradle(task: "jacocoRootReport --no-daemon")
-    gradle(task: "sample:assembleDebug --no-daemon")
-    gradle(task: "sonarqube --no-daemon")
+    gradle(task: "clean")
+    gradle(task: "detekt")
+    gradle(task: "beagle:lintDebug")
```

Comment:
this will disable the debug link to other modules, is it really correct?

Reply:
Lint tends to get slower for every release. There's a reason for that: It keeps checking more and more things (as of 3.3 canary we're up to around 350 separate issues), and existing issues are often made more complex as well, doing deeper flow and data analysis, and occasionally there are new features which also adds cost, such as baselines, Kotlin analysis, etc. And this is all compounded by the codebases lint is getting run on also growing bigger and bigger, and being broken up into more and more modules.

A number of you have complained about the performance and asked how to make it better, so I figured it would be helpful to try to write a "performance tuning guide" with tips. There are several things you can do in your setup to run lint a LOT faster.

This is a draft, and I'd like your feedback and results on this, and then one of our doc writers can take the raw material and turn it into real documentation.

-----------------------------------------
Use lintRelease target, not lint
-----------------------------------------

If you run "./gradlew tasks" you'll see that there's a "lint" task, and you may be tempted to run it. But lint also adds specific tasks for each variant. For example, if you only have "debug" and "release" variants (e.g. you haven't defined any product flavors or additional build types), lint will create 3 targets:
"lint"
"lintDebug"
"lintRelease"

Don't configure your CI job to run the "lint" task; pick a specific variant instead, such as "lintDebug". The reason this is so important is that what the "lint" task really does is run lint over and over again, for *each* variant, and then it collates the results in the end; combines all the errors it found, and shows which specific variants every error applies to, unless it applies to all (which is commonly the case).   The reason for this is that you could have a variant-specific issue; for example, you may have a resource file which is only present in a variant-specific source set (such as src/debug/res/values), so the general lint target checks everything. Of course, you're probably only shipping your release variant, so you could limit yourself to just running "lintRelease" and you're not going to miss much.

If you have lots of variants this makes a huge difference; with 7 variants, "lint" will take ~7x longer than "lintDebug" !!  (modulo some caching)

---------------------------------------
Only analyze app/leaf modules
---------------------------------------

If you have divided your project into many smaller modules - a number of libraries and just a couple of app modules, it's much better to (a) turn on checkDependencies mode and (b) only run lint on the app modules, instead of recursively running lint on each module.

To do this, first add this to your app module's build.gradle file:
```
android {
    ...
    lintOptions {
        ...
        checkDependencies true
        ...
    }
}         
```

Then instead of running ./gradlew lintDebug, run ./gradlew :app:lintDebug.

Since you've turned on check-dependencies-mode, running lint on the app module will also run it on all the dependent modules the app depends on -- e.g. all the libraries. But this should also make things run a lot faster, since it's a single lint invocation, where lint shares a lot more computation (such as symbol resolution in the SDK and various shared libraries, etc).  

This isn't just a good idea from a performance perspective; you'll also get more accurate results. For example, the unused resource analysis will now correctly know whether a resource defined in a library is consumed by the app module; that doesn't happen when you run lint first on the library, and later on just the app, which is what happens when you run ./gradlew lintDebug. As a bonus you'll also get a single HTML report containing all the results from your codebase, instead of having to hunt around the tree for all the various reports and look at each one separately.

---------------------------------------
Don't analyze test sources
---------------------------------------

Lint has two flags controlling what to do about test sources: checkTestSources and ignoreTestSources. These mean different things.

"checkTestSources" controls whether lint should run all the normal checks on the test sources. This is already off by default, so there's really no performance reason to tweak it. The only reason you'd turn this on is if you really want to make sure your test sources are pristine as well.
"ignoreTestSources" on the other hand controls whether lint should really ignore all test sources. Even if lint doesn't run normal checks on the test sources, it still has to include them in analysis for some of the regular lint checks for production code. For example, what if you have a resource which is only referenced from a test; do you then want to flag this as unused when the test is still referencing it?

Add this to your app build.gradle file to tell lint to completely ignore your test sources, which should help lint run faster since (for well tested code) this can help skip a lot of code analysis and type resolution:
```
android {
    ...
    lintOptions {
        ...
        ignoreTestSources true
        ...
    }
}         
```